src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs

```diff
@@ -402,14 +402,23 @@ public override void VisitForStatement(ForStatementSyntax node)
                 } 
             }
 
-            if (node.Condition != null)
+            ExpressionSyntax condition = node.Condition;
+            if (condition != null)
             {
-                Visit(node.Condition, binder);
+                binder = new ExpressionVariableBinder(condition, binder);
+                AddToMap(condition, binder);
+                Visit(condition, binder);
             }
 
-            foreach (var incrementor in node.Incrementors)
+            SeparatedSyntaxList<ExpressionSyntax> incrementors = node.Incrementors;
+            if (incrementors.Count > 0)
             {
-                Visit(incrementor, binder);
+                var incrementorsBinder = new ExpressionListVariableBinder(incrementors, binder);
```

Comment:
Do you plan to use `ExpressionListVariableBinder` for other purposes than binding incrementors? If not, consider calling it `IncrementorsBinder`